x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,367 @@
+    let path = get_path();
+
+    let Child { stdin, stdout, .. } = Command::new(path)
+        .arg("interpreter")
+        .arg("--cleanup")
+        .arg("--log-level")
+        .arg("debug")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let writer = stdin.ok_or(ClientError::StdIo)?;
+    let reader = stdout.ok_or(ClientError::StdIo)?;
+
+    let responses = BufReader::new(reader)
+        .lines()
+        .map(|line| serde_json::from_str(&line?).map_err(SimulatorError::Serde));
+
+    Ok(Simulator { writer, responses })
+}
+
+impl<W, R> Simulator<W, R>
+where
+    W: Write,
+    R: Iterator<Item = SimulatorResponseItem>,
+{
+    const RUN_COMMAND: &'static [u8] = b"run --message '";
+
+    pub fn create_program<P: AsRef<Path>>(&mut self, path: P) -> SimulatorResponseItem {
+        let path = path.as_ref().to_string_lossy();
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::new_create_program(path.into()))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|resp: SimulatorResponse| {
+                if let Some(err) = resp.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(resp)
+                }
+            })
+    }
+
+    pub fn read(&mut self, method: String, params: Vec<Param>) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::new_read(method, params))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|request| {
+                if let Some(err) = request.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(request)
+                }
+            })
+    }
+
+    pub fn execute(
+        &mut self,
+        method: String,
+        params: Vec<Param>,
+        max_units: u64,
+    ) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::new_execute(method, params, max_units))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|request| {
+                if let Some(err) = request.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(request)
```

Comment:
Same comment as above

Reply:
Also, the third time I'm seeing the exact same logic. Duplicating twice is fine, but three times deserves an abstraction. 

```rust
impl From<SimulatorResponse> for Result<SimulatorResponse, SimulatorError> {
// ...
}
```